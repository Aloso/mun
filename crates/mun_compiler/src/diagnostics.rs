use mun_hir::diagnostics::{Diagnostic as HirDiagnostic, DiagnosticSink};
use mun_hir::{FileId, HirDatabase, HirDisplay, Module};
use mun_syntax::{ast, AstNode, SyntaxKind};

use std::cell::RefCell;

use crate::annotate_snippets_builders::{AnnotationBuilder, SliceBuilder, SnippetBuilder};

use annotate_snippets::snippet::{AnnotationType, Snippet};

/// Constructs diagnostic messages for the given file.
pub fn diagnostics(db: &impl HirDatabase, file_id: FileId) -> Vec<Snippet> {
    let parse = db.parse(file_id);

    let mut result = Vec::new();

    let source_code = db.file_text(file_id).to_string().replace("\t", " ");
    let source_code_len = source_code.len();

    let relative_file_path = db.file_relative_path(file_id).display().to_string();

    let line_index = db.line_index(file_id);

    result.extend(parse.errors().iter().map(|err| {
        let mut snippet = SnippetBuilder::new()
            .title(
                AnnotationBuilder::new(AnnotationType::Error)
                    .label("syntax error".to_string())
                    .build(),
            )
            .slice(
                SliceBuilder::new(true)
                    .origin(relative_file_path.clone())
                    .source_annotation(
                        (
                            err.location().offset().to_usize(),
                            err.location().end_offset().to_usize(),
                        ),
                        err.to_string(),
                        AnnotationType::Error,
                    )
                    .build(&source_code, source_code_len, &line_index),
            )
            .build();
        // Add one to right range to make highlighting range here visible on output
        snippet.slices[0].annotations[0].range.1 += 1;

        snippet
    }));

    let result = RefCell::new(result);
    let mut sink = DiagnosticSink::new(|d| {
        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label("unexpected error".to_string())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            (
                                d.highlight_range().start().to_usize(),
                                d.highlight_range().end().to_usize(),
                            ),
                            d.message(),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::UnresolvedValue, _>(|d| {
        let unresolved_value = d.expr.to_node(&parse.tree().syntax()).text().to_string();

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(format!(
                            "cannot find value `{}` in this scope",
                            unresolved_value
                        ))
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            (
                                d.highlight_range().start().to_usize(),
                                d.highlight_range().end().to_usize(),
                            ),
                            "not found in this scope".to_string(),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::UnresolvedType, _>(|d| {
        let unresolved_type = d
            .type_ref
            .to_node(&parse.syntax_node())
            .syntax()
            .text()
            .to_string();

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(d.message())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            (
                                d.highlight_range().start().to_usize(),
                                d.highlight_range().end().to_usize(),
                            ),
                            format!("could not find type `{}` in this scope", unresolved_type),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::ExpectedFunction, _>(|d| {
        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(d.message())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            (
                                d.highlight_range().start().to_usize(),
                                d.highlight_range().end().to_usize(),
                            ),
                            format!("expected function, found `{}`", d.found.display(db)),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::MismatchedType, _>(|d| {
        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(d.message())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            (
                                d.highlight_range().start().to_usize(),
                                d.highlight_range().end().to_usize(),
                            ),
                            format!(
                                "expected `{}`, found `{}`",
                                d.expected.display(db),
                                d.found.display(db)
                            ),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::DuplicateDefinition, _>(|d| {
        let first_definition_location = match d.first_definition.kind() {
            SyntaxKind::FUNCTION_DEF => {
                ast::FunctionDef::cast(d.first_definition.to_node(&parse.tree().syntax()))
                    .map(|f| f.signature_range())
                    .unwrap_or_else(|| d.first_definition.range())
            }
            SyntaxKind::STRUCT_DEF => {
                ast::StructDef::cast(d.first_definition.to_node(&parse.tree().syntax()))
                    .map(|s| s.signature_range())
                    .unwrap_or_else(|| d.first_definition.range())
            }
            _ => d.first_definition.range(),
        };
        let definition_location = match d.definition.kind() {
            SyntaxKind::FUNCTION_DEF => {
                ast::FunctionDef::cast(d.definition.to_node(&parse.tree().syntax()))
                    .map(|f| f.signature_range())
                    .unwrap_or_else(|| d.definition.range())
            }
            SyntaxKind::STRUCT_DEF => {
                ast::StructDef::cast(d.definition.to_node(&parse.tree().syntax()))
                    .map(|s| s.signature_range())
                    .unwrap_or_else(|| d.definition.range())
            }
            _ => d.definition.range(),
        };
        let duplication_object_type = if matches!(d.first_definition.kind(), SyntaxKind::STRUCT_DEF)
            && matches!(d.definition.kind(), SyntaxKind::STRUCT_DEF)
        {
            "type"
        } else {
            "value"
        };
        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(d.message())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        // First definition
                        .source_annotation(
                            (
                                first_definition_location.start().to_usize(),
                                first_definition_location.end().to_usize(),
                            ),
                            format!(
                                "previous definition of the {} `{}` here",
                                duplication_object_type, d.name
                            ),
                            AnnotationType::Warning,
                        )
                        // Second definition
                        .source_annotation(
                            (
                                definition_location.start().to_usize(),
                                definition_location.end().to_usize(),
                            ),
                            format!("`{}` redefined here", d.name),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .footer(
                    AnnotationBuilder::new(AnnotationType::Note)
                        .label(format!(
                            "`{}` must be defined only once in the {} namespace of this module",
                            d.name, duplication_object_type
                        ))
                        .build(),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::PossiblyUninitializedVariable, _>(|d| {
        let variable_name = d.pat.to_node(&parse.syntax_node()).text();

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(format!("{}: `{}`", d.message(), variable_name))
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            (
                                d.highlight_range().start().to_usize(),
                                d.highlight_range().end().to_usize(),
                            ),
                            format!("use of possibly-uninitialized `{}`", variable_name),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::AccessUnknownField, _>(|d| {
        let location = ast::FieldExpr::cast(d.expr.to_node(&parse.syntax_node()))
            .map(|f| f.field_range())
            .unwrap_or_else(|| d.highlight_range());

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(format!(
                            "no field `{}` on type `{}`",
                            d.name,
                            d.receiver_ty.display(db),
                        ))
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            (location.start().to_usize(), location.end().to_usize()),
                            "unknown field".to_string(),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    });

    Module::from(file_id).diagnostics(db, &mut sink);

    drop(sink);

    result.into_inner()
}
