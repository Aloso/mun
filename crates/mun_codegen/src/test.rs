use crate::{mock::MockDatabase, IrDatabase};
use mun_hir::diagnostics::DiagnosticSink;
use mun_hir::Module;
use std::cell::RefCell;
use std::path::PathBuf;
use test_utils::{dir_tests, project_dir};

fn test_data_dir() -> PathBuf {
    project_dir().join("crates/mun_codegen/tests/data/")
}

#[test]
fn ir_tests() {
    dir_tests(&test_data_dir(), &["ir"], |text, _path| {
        let (db, file_id) = MockDatabase::with_single_file(text);

        let messages = RefCell::new(Vec::new());
        let mut sink = DiagnosticSink::new(|diag| {
            messages.borrow_mut().push(diag.message());
        });
        if let Some(module) = Module::package_modules(&db)
            .iter()
            .find(|m| m.file_id() == file_id)
        {
            module.diagnostics(&db, &mut sink)
        }
        drop(sink);
        let messages = messages.into_inner();

        if !messages.is_empty() {
            messages.join("\n")
        } else {
            format!(
                "{}",
                db.module_ir(file_id)
                    .llvm_module
                    .print_to_string()
                    .to_string()
            )
        }
    });
}
