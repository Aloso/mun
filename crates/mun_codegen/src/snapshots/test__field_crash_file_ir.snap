---
source: crates/mun_codegen/src/test.rs
expression: "struct(gc) Foo { a: int };\n\nfn main(c:int):int {\n    let b = Foo { a: c + 5 }\n    b.a\n}"
---
; ModuleID = 'main.mun'
source_filename = "main.mun"

%DispatchTable = type { i8* addrspace(4)* (i8 addrspace(4)*, i8*)* }
%struct.MunTypeInfo = type { [16 x i8], i8 addrspace(4)*, i32, i8, i8 }
%Foo = type { i64 }

@allocatorHandle = external global i8*
@dispatchTable = external global %DispatchTable
@global_type_table = external global [5 x %struct.MunTypeInfo addrspace(4)*]

define i64 @main(i64) {
body:
  %b = alloca %Foo* addrspace(4)*
  %c = alloca i64
  store i64 %0, i64* %c
  %c1 = load i64, i64* %c
  %add = add i64 %c1, 5
  %init = insertvalue %Foo undef, i64 %add, 0
  %new_ptr = load i8* addrspace(4)* (i8 addrspace(4)*, i8*)*, i8* addrspace(4)* (i8 addrspace(4)*, i8*)** getelementptr inbounds (%DispatchTable, %DispatchTable* @dispatchTable, i32 0, i32 0)
  %Foo_ptr = load %struct.MunTypeInfo addrspace(4)*, %struct.MunTypeInfo addrspace(4)** getelementptr inbounds ([5 x %struct.MunTypeInfo addrspace(4)*], [5 x %struct.MunTypeInfo addrspace(4)*]* @global_type_table, i32 0, i32 0)
  %type_info_ptr_to_i8_ptr = bitcast %struct.MunTypeInfo addrspace(4)* %Foo_ptr to i8 addrspace(4)*
  %allocator_handle = load i8*, i8** @allocatorHandle
  %new = call i8* addrspace(4)* %new_ptr(i8 addrspace(4)* %type_info_ptr_to_i8_ptr, i8* %allocator_handle)
  %Foo_ptr_ptr = bitcast i8* addrspace(4)* %new to %Foo* addrspace(4)*
  %Foo_mem_ptr = load %Foo*, %Foo* addrspace(4)* %Foo_ptr_ptr
  store %Foo %init, %Foo* %Foo_mem_ptr
  store %Foo* addrspace(4)* %Foo_ptr_ptr, %Foo* addrspace(4)** %b
  %mem_ptr = load %Foo* addrspace(4)*, %Foo* addrspace(4)** %b
  %deref = load %Foo*, %Foo* addrspace(4)* %mem_ptr
  %Foo.a = getelementptr inbounds %Foo, %Foo* %deref, i32 0, i32 0
  %a = load i64, i64* %Foo.a
  ret i64 %a
}

